### ISSUE TITLE
urljoin works incorrectly for two path-relative URLs involving . and ..

### ISSUE THREAD

**Issue by andersk** [python/cpython#96015]
# Bug report

`urllib.parse.urljoin` is usually used to join a normalized absolute URL with a relative URL, and it generally works for that purpose. But if it’s used to join two path-relative URLs, it produces incorrect results in many cases when `.` or `..` is involved.

```pycon
>>> from urllib.parse import urljoin
>>> urljoin('a', 'b')  # ok
'b'
>>> urljoin('a/', 'b')  # ok
'a/b'
>>> urljoin('a', '.')  # expected . or ./
'/'
>>> urljoin('a', '..')  # expected .. or ../
'/'
>>> urljoin('..', 'b')  # expected ../b
'b'
>>> urljoin('../a', 'b')  # expected ../b
'b'
>>> urljoin('a', '../b')  # expected ../b
'b'
>>> urljoin('../a', '../b')  # expected ../../b
'b'
>>> urljoin('a/..', 'b')  # expected b
'a/b'
```

There are also some problems when the base is a non-normalized absolute URL:

```pycon
>>> urljoin('http://host/a/..', 'b')  # expected http://host/b
'http://host/a/b'
```

# Your environment

- CPython versions tested on: 3.10.5, 3.11.0rc1
- Operating system and architecture: NixOS 21.11 amd64

**Comment by 6t8k** [comment]
According to the docs on [`urllib.urlparse.urljoin`](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urljoin) and the [docstring](https://github.com/python/cpython/blob/3.11/Lib/urllib/parse.py#L514) in its code, the function is meant to construct an absolute URL, so I think that's what should be expected from its return value.

At least as far as [RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986.html#section-5.1) is concerned, which is mentioned in the linked docs, the "Base URI" is also expected to be absolute. Maybe the docs could be a bit clearer.

**Comment by andersk** [comment]
Even that excuse doesn’t apply to `urljoin('http://host/a/..', 'b')`.

**Comment by 6t8k** [comment]
That indeed appears to be a more subtle case. Some clues, without a conclusion from me:

- RFC 3986 sections [3.3](https://datatracker.ietf.org/doc/html/rfc3986.html#section-3.3) and [6.2.2.3](https://datatracker.ietf.org/doc/html/rfc3986.html#section-6.2.2.3) say dot-segments are intended for use at the beginning of a relative-path reference
- [5.2.1](https://datatracker.ietf.org/doc/html/rfc3986.html#section-5.2.1) says normalization of the base URI, as described in [6.2.2](https://datatracker.ietf.org/doc/html/rfc3986.html#section-6.2.2) and [6.2.3](https://datatracker.ietf.org/doc/html/rfc3986.html#section-6.2.3), is _optional_ (for historical/compatibility reasons, I'd assume)
- In `urllib.urlparse.urljoin`, the last segment of the base URI, here a dot-segment, is removed [here](https://github.com/python/cpython/blob/3.11/Lib/urllib/parse.py#L547) (this is not using the `remove_dot_segments` algorithm from [5.2.4](https://datatracker.ietf.org/doc/html/rfc3986.html#section-5.2.4) – effectively the segment is ignored)
- The [`URL` API](https://developer.mozilla.org/en-US/docs/Web/API/URL/URL), in major web browsers, does produce `http://host/b` (but for relative bases, it throws errors). It seems that they opt to not skip normailzation of the base URI; the [`new URL(url, base)`](https://url.spec.whatwg.org/#dom-url-url) constructor's, and following this, [`basic URL parser`](https://url.spec.whatwg.org/#concept-basic-url-parser)'s [`path state`](https://url.spec.whatwg.org/#path-state) specs seem to indicate this:
<details>

Firefox 107.0

```javascript
let baseUrl1 = "http://host/a/..";
undefined
let A = new URL("b", baseUrl1);
undefined
A.href
"http://host/b"
let baseUrl2 = "../a";
undefined
let B = new URL("b", baseUrl2);
Uncaught TypeError: URL constructor: ../a is not a valid URL.
    <anonymous> debugger eval code:1
debugger eval code:1:9 
```

Chromium 107.0.5304.121
```javascript
let baseUrl1 = "http://host/a/..";
undefined
let A = new URL("b", baseUrl1);
undefined
A.href
'http://host/b'
let baseUrl2 = "../a";
undefined
let B = new URL("b", baseUrl2);
VM520:1 Uncaught TypeError: Failed to construct 'URL': Invalid base URL
    at <anonymous>:1:15
(anonymous) @ VM520:1
```
  </details>

---

The question then is, I suppose, whether this behavior is unexpected/hard to work around enough that fixing it would be worthwhile to the point of making the risk of breaking existing code acceptable. So far I'm not aware of any standards-breaking behavior, at least.

**Comment by andersk** [comment]
The expected semantics are obvious. These URLs are equivalent:

`"http://host/a/.." ~ "http://host/"`

so we expect these URLs to be equvalent:

`urljoin("http://host/a/..", "b") ~ urljoin("http://host/", "b")`.

In general, we always expect:

- if `a ~ a1` and `b ~ b1`, then `urljoin(a, b) ~ urljoin(a1, b1)`

(where `a ~ a1` is the equivalence relation `normalize(a) == normalize(a1)`).

Any time equivalent inputs lead to nonequivalent outputs is an opportunity for bugs. If the standard makes it optional to produce an expected result for legacy reasons, that doesn’t make it a good idea to produce an unexpected result.

---

Another law we always expect is

- `urljoin(a, urljoin(b, c)) ~ urljoin(urljoin(a, b), c)`

which makes it obvious how `urljoin(b, c)` is expected to behave for relative `b` (if it’s not going to throw an error). For example, we should have

```
urljoin("http://host/w/x/y/z", urljoin("../b", "../c"))
~ urljoin(urljoin("http://host/w/x/y/z", "../b"), "../c")
== urljoin("http://host/w/x/b", "../c")
== "http://host/w/c"
== urljoin("http://host/w/x/y/z", "../../c")
```

which can only work if `urljoin("../b", "../c") ~ "../../c"`.

**Comment by andersk** [comment]
@serhiy-storchaka The RFC 3986 algorithm is defined for an absolute base URI and a relative-path reference. It *does not make sense* to indiscriminately apply that same algorithm to two relative-path references. [§5.1](https://datatracker.ietf.org/doc/html/rfc3986#section-5.1):

> A base URI must conform to the `<absolute-URI>` syntax rule ([Section 4.3](https://datatracker.ietf.org/doc/html/rfc3986#section-4.3)).

[§5.2.1](https://datatracker.ietf.org/doc/html/rfc3986#section-5.2.1):

> Note that only the scheme component is required to be present in a base URI; the other components may be empty or undefined.

If we are to insist on using the algorithm as written, we’d have to throw an error when the base URI is missing a scheme.

However, it does make sense, and is useful in practice, to specify a generalized `urljoin(a, b)` operation that aligns with the RFC operation when `a` *is* an absolute URI, and that also satisfies the expected properties I listed [above](https://github.com/python/cpython/issues/96015#issuecomment-1327921527). We do this by preserving extra initial `..` components in the output when given non–RFC 3986 input where the base URI is path-relative (undefined scheme, undefined authority, and path not beginning with `/`).

**Comment by serhiy-storchaka** [comment]
An absolute base URI can have an undefined authority and a rootless or empty path.

**Comment by andersk** [comment]
Yes it can, if its scheme is defined and the parser is “strict” in the RFC sense. If we intend to behave as a strict parser, we use the RFC algorithm for that case and there is no associativity issue.

A non-strict parser considers such a URI to be a relative reference. [§5.4.2](https://datatracker.ietf.org/doc/html/rfc3986#section-5.4.2):

> Some parsers allow the scheme name to be present in a relative reference if it is the same as the base URI scheme.  This is considered to be a loophole in prior specifications of partial URI [[RFC1630](https://datatracker.ietf.org/doc/html/rfc1630)].  Its use should be avoided but is allowed for backward compatibility.
>
>     "http:g"        =  "http:g"         ; for strict parsers
>                     /  "http://a/b/c/g" ; for backward compatibility

[§5.2.2](https://datatracker.ietf.org/doc/html/rfc3986#section-5.2.2):

>     -- A non-strict parser may ignore a scheme in the reference
>     -- if it is identical to the base URI's scheme.
>     --
>     if ((not strict) and (R.scheme == Base.scheme)) then
>        undefine(R.scheme);
>     endif;

If we intend to behave as a non-strict parser, then `urljoin('http:../a', 'http:../b')` again falls outside the scope of the RFC, but associativity mandates that we should preserve initial `..` components and return `http:../../b`, if we’re being totally pedantic about our non-strictness. (I’m not too concerned about this obscure case.)

**Comment by picnixz** [comment]
From https://github.com/python/cpython/issues/96015#issuecomment-2468075398, I'm considering we have a bit of bugs and feature requests but anything that fixes this would still be breaking existing code (although some could argue that "garbage in -> garbage out" in some cases). So I'm still marking it as a FR but if we want to track what can be really considered as a bugs, it'd be good. Nonetheless,

> However, it does make sense, and is useful in practice, to specify a generalized urljoin(a, b) operation that aligns with the RFC operation when a is an absolute URI, and that also satisfies the expected properties I listed https://github.com/python/cpython/issues/96015#issuecomment-1327921527. We do this by preserving extra initial .. components in the output when given non–RFC 3986 input where the base URI is path-relative (undefined scheme, undefined authority, and path not beginning with /).

makes me consider this request as a FR rather than a bugfix.

**Comment by ericvsmith** [comment]
It's definitely not a bug fix: there's no way we'd backport this, even in the unlikely event it were to be accepted.
